cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -595,30 +620,50 @@ func (m *stateMachine) setUnlockStateWithFallbackKey(partName string, unlockRes
+	// we can't use the run object keys from there and instead need to directly
+	// fallback to trying the fallback object keys from ubuntu-seed
+	// - The second case is if we couldn't identify an ubuntu-data-enc or an
+	// ubuntu-data partition at all, we still could have an ubuntu-save-enc
+	// partition in which case we maybe could still have an encrypted disk that
+	// needs unlocking with the fallback object keys from ubuntu-seed
+	//
+	// As such, if m.isEncryptedDev is false, but unlockRes.IsDecryptedDevice is
+	// true, then it is safe to assign m.isEncryptedDev to true.
 	if !m.isEncryptedDev && unlockRes.IsDecryptedDevice {
 		m.isEncryptedDev = unlockRes.IsDecryptedDevice
 	}
 
-	// also make sure that if we previously saw a device that we see the same
-	// device again
-	if unlockRes.Device != "" && part.Device != "" && unlockRes.Device != part.Device {
-		return fmt.Errorf("inconsistent partitions found for %s: previously found %s but now found %s", partName, part.Device, unlockRes.Device)
+	// Also make sure that if we previously saw a device that we see the same
+	// device again.
+	if unlockRes.Device != "" && part.rawDevice != "" && unlockRes.Device != part.rawDevice {
+		return fmt.Errorf("inconsistent partitions found for %s: previously found %s but now found %s", partName, part.rawDevice, unlockRes.Device)
 	}
 
-	if unlockRes.Device != "" {
-		part.FindState = partitionFound
-		part.Device = unlockRes.Device
+	// Note that we don't do any consistency checking on part.decryptedRawDevice
+	// and unlockRes.DecryptedDevice because unlockRes.DecryptedDevice should
+	// only have been set if the decryption result succeeded, so
+	// we should never see multiple UnlockResult's with DecryptedDevice. This is
+	// because the first time that it succeeded with the run object key we would not have entered into
+	// the fallback case here.
+
+	// We do however do consistency checking on unlockRes to make sure the
+	// result makes sense.
+	if !unlockRes.IsDecryptedDevice && unlockRes.DecryptedDevice != "" && err != nil {
+		// This case should be impossible to enter, if the device is not
+		// decrypted, then what did we unlock to get a decrypted device?
+		return fmt.Errorf("internal error: inconsistent return values from UnlockVolumeUsingSealedKeyIfEncrypted for partition %s: %v", partName, err)
 	}
 
-	if !unlockRes.IsDecryptedDevice && unlockRes.Device != "" && err != nil {
-		// this case should be impossible to enter, if we have an unencrypted
-		// device and we know what the device is then what is the error?
-		return fmt.Errorf("internal error: inconsistent return values from UnlockVolumeUsingSealedKeyIfEncrypted for partition %s", partName)
+	if unlockRes.Device != "" {
+		part.FindState = partitionFound
+		// Note that in some case this may be redundantly assigning the same
+		// value to rawDevice again.
+		part.rawDevice = unlockRes.Device
+		part.decryptRawDevice = unlockRes.Device
```

Comment:
again this should be set from DecryptedDevice

Reply:
this now uses FsDevice